deep_qa/common/params.py

```diff
@@ -1,52 +1,184 @@
-from typing import Any, Dict, List
+from typing import Any, Dict, List, Union
+from collections import MutableMapping
 
+import logging
 import pyhocon
 
+from overrides import overrides
 from .checks import ConfigurationError
 
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
 
-def _get_choice_error_message(value: Any, choices: List[Any], name: str=None) -> str:
-    if name:
-        return '%s not in acceptable choices for %s: %s' % (value, name, str(choices))
-    else:
-        return '%s not in acceptable choices: %s' % (value, str(choices))
+PARAMETER = 60
+logging.addLevelName(PARAMETER, "PARAM")
 
 
-def get_choice(params: Dict[str, Any], key: str, choices: List[Any], name: str=None):
+def __param(self, message, *args, **kws):
     """
-    Gets the value of `key` in the `params` dictionary, ensuring that the value is one of the given
-    choices.  `name` is an optional description for where a configuration error happened, if there
-    is one.
+    Add a method to logger which allows us
+    to always log parameters unless you set the logging
+    level to be higher than 60 (which is higher than the
+    standard highest level of 50, corresponding to WARNING).
+    """
+    # Logger takes its '*args' as 'args'.
+    if self.isEnabledFor(PARAMETER):
+        self._log(PARAMETER, message, args, **kws) # pylint: disable=protected-access
+logging.Logger.param = __param
+
 
-    Note that this _pops_ the key from params, modifying the dictionary, consistent with how
-    parameters are processed in this codebase.
+class Params(MutableMapping):
     """
-    value = params.pop(key)
-    if value not in choices:
-        raise ConfigurationError(_get_choice_error_message(value, choices, name))
-    return value
+    A Class representing a parameter dictionary with a history. Using this allows
+    exact reproduction of a parameter file used in an experiment from logs, even when
+    default values are used.
+    """
+
+    # This allows us to check for the presence of "None" as a default argument,
+    # which we require because we make a distinction bewteen passing a value of "None"
+    # and passing no value to the default parameter of "pop".
+    DEFAULT = object()
+
+    def __init__(self, params: Dict[str, Any], history: str=""):
+
+        self.params = params
+        self.history = history
+
+    @overrides
+    def pop(self, key: str, default: Any=DEFAULT):
+        """
+        Performs the functionality associated with dict.pop(key), along with checking for
+        returned dictionaries, replacing them with Param objects with an updated history.
+        """
+        if default is self.DEFAULT:
+            value = self.params.pop(key)
+        else:
+            value = self.params.pop(key, default)
+        logger.param(self.history + " : " + key + " : " + str(value))
```

Comment:
I'd probably change this format slightly, here and below.  If you just do `history + "." + key`, this is directly readable with pyhocon, because, e.g., `embeddings_dim.words = 100` is valid hocon.  Doing it this way also suggests that you can just have `history` always end with a `.`, so you only need to do `history + key` here, and it will just work when `history` is empty, too.